File: worker/src/figma.ts
/**
 * The Figma side: read a frame, and pin things back onto it.
 *
 * A port of bridge/src/bridge/figma.py, kept in the same shape so the two can be read side by
 * side. The three facts that decide the design are unchanged, because they are facts about
 * Figma rather than about either runtime:
 *
 * REST cannot create or change a node. Not a frame, not a layer, not a text run. So this
 * credential, whatever scopes it carries, cannot damage a designer's file. The worst it can do
 * is leave a link and a comment.
 *
 * REST cannot set a frame's dev status either. Marking a frame completed stays a person's click,
 * which is the right place for it.
 *
 * A frame names components by Figma component key, and the pack names them by code import. The
 * key map joins the two, and what it cannot join is reported as unmapped rather than guessed at.
 */
const API = 'https://api.figma.com'

/** Figma reports fractional sizes. Round, and treat a missing or zero value as unknown. */
const round = (value: number | undefined): number | null =>
  typeof value === 'number' && value > 0 ? Math.round(value) : null

/**
 * A component in the frame that the key map could not place.
 *
 * Carries the size the design gave it, so a placeholder can reserve the same space rather than
 * collapsing the layout around a hole. `null` where Figma did not report one.
 */
export interface UnmappedComponent {
  name: string
  width: number | null
  height: number | null
  radius: number | null
}

export interface FrameRead {
  /**
   * How the frame is arranged, top to bottom, already rendered as markdown list lines.
   *
   * The component list is a SET, sorted, so it says what the frame needs and nothing about where
   * anything sits. An agent handed only that has to invent an arrangement, and it did: it put a
   * placeholder above the figures the design puts it below. The tree is walked in order anyway,
   * so the ordering existed and was being thrown away.
   */
  outline: string[]
  nodeId: string
  name: string
  pageName: string
  fileKey: string
  fileName: string
  components: string[]
  unknownComponents: UnmappedComponent[]
  text: string[]
}

/** The link a person opens. Figma writes 41-207 in a URL and 41:207 in the API. */
export function frameUrl(fileKey: string, nodeId: string): string {
  return `https://www.figma.com/design/${fileKey}/?node-id=${nodeId.replace(':', '-')}`
}

interface Node {
  id?: string
  name?: string
  type?: string
  characters?: string
  componentId?: string
  layoutMode?: string
  cornerRadius?: number
  absoluteBoundingBox?: { width?: number; height?: number }
  children?: Node[]
}

/** A quoted, shortened line of copy. Long strings make an outline unreadable. */
const quoted = (text: string): string => `"${text.length > 60 ? `${text.slice(0, 57)}...` : text}"`

/**
 * The frame's arrangement, in document order, as markdown list lines.
 *
 * Instances are opaque. The walk does not descend into one, mapped or not, because what is inside
 * a component is the component's business and listing it invites an agent to rebuild it.
 *
 * A frame that lays its children out becomes a level and says which way it runs. A frame that does
 * not is transparent: it nests nothing and its children are emitted where it sat, because a
 * grouping with no layout is a drawing convenience rather than structure.
 *
 * Adjacent identical entries collapse to `x N`. Three figures in a row is one fact, not three.
 */
function outlineOf(
  root: Node,
  meta: Record<string, { key?: string; name?: string }>,
  packComponents: Record<string, string>,
  depth = 0,
): string[] {
  if (depth > 3) return []
  const lines: string[] = []
  const pad = '  '.repeat(depth)

  const entries: string[] = []
  const nested: Record<number, string[]> = {}

  for (const child of root.children ?? []) {
    if (child.type === 'TEXT' && child.characters) {
      entries.push(`${pad}- ${quoted(child.characters)}`)
      continue
    }
    if (child.type === 'INSTANCE') {
      const key = meta[child.componentId ?? '']?.key ?? ''
      const mapped = packComponents[key]
      const name = mapped ?? meta[child.componentId ?? '']?.name ?? child.name ?? 'unnamed'
      entries.push(`${pad}- ${mapped ? `\`${name}\`` : `**${name}**`}`)
      continue
    }
    if (child.children?.length) {
      const inner = child.layoutMode === 'HORIZONTAL' || child.layoutMode === 'VERTICAL'
      const label = child.layoutMode === 'HORIZONTAL' ? 'a row of:' : 'a column of:'
      if (inner) {
        entries.push(`${pad}- ${label}`)
        nested[entries.length - 1] = outlineOf(child, meta, packComponents, depth + 1)
      } else {
        // Transparent. A grouping with no layout is a drawing convenience, not structure.
        for (const line of outlineOf(child, meta, packComponents, depth)) entries.push(line)
      }
    }
  }

  // Collapse runs of the same entry. Three of a thing in a row is one fact.
  let index = 0
  while (index < entries.length) {
    let run = 1
    while (
      index + run < entries.length &&
      entries[index + run] === entries[index] &&
      nested[index] === undefined &&
      nested[index + run] === undefined
    ) {
      run += 1
    }
    lines.push(run > 1 ? `${entries[index]} x ${run}` : entries[index])
    for (const line of nested[index] ?? []) lines.push(line)
    index += run
  }
  return lines
}

export class FigmaClient {
  constructor(
    private readonly token: string,
    /**
     * Wrapped, not `= fetch`. Stored as a property and called as `this.fetcher(...)`, the global
     * `fetch` receives this client as its `this`, and workerd refuses that with "Illegal
     * invocation". Node tolerates it, so the whole suite passed and the deployed Worker threw a
     * 500 on the first real webhook. The arrow keeps the call a plain global call.
     */
    private readonly fetcher: typeof fetch = (input, init) => fetch(input, init),
  ) {}

  private headers(extra: Record<string, string> = {}): Record<string, string> {
    return { 'X-Figma-Token': this.token, ...extra }
  }

  private async get(path: string, params: Record<string, string> = {}): Promise<any> {
    const url = new URL(API + path)
    for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value)
    const response = await this.fetcher(url.toString(), { headers: this.headers() })
    if (!response.ok) throw new Error(`Figma GET ${path} -> ${response.status}`)
    return response.json()
  }

  private async post(path: string, payload: unknown): Promise<any> {
    const response = await this.fetcher(API + path, {
      method: 'POST',
      headers: this.headers({ 'Content-Type': 'application/json' }),
      body: JSON.stringify(payload),
    })
    if (!response.ok) throw new Error(`Figma POST ${path} -> ${response.status}`)
    return response.json()
  }

  /**
   * Read one frame and resolve what it is made of against the pack.
   *
   * `packComponents` maps a Figma component key to a pack component name. A component in the
   * frame that is not in that map goes into `unknownComponents`, because an agent told the wrong
   * component name confidently is worse than one told the truth vaguely.
   */
  async readFrame(
    fileKey: string,
    nodeId: string,
    packComponents: Record<string, string>,
  ): Promise<FrameRead> {
    const data = await this.get(`/v1/files/${fileKey}/nodes`, { ids: nodeId, depth: '8' })
    const entry = data?.nodes?.[nodeId]
    if (!entry) throw new Error(`node ${nodeId} is not in file ${fileKey}`)

    const meta: Record<string, { key?: string; name?: string }> = entry.components ?? {}
    const known = new Set<string>()
    const unknown = new Map<string, UnmappedComponent>()
    const text: string[] = []

    const visit = (node: Node): void => {
      if (node.type === 'INSTANCE') {
        const key = meta[node.componentId ?? '']?.key ?? ''
        const name = packComponents[key]
        if (name) {
          known.add(name)
        } else {
          const missing = meta[node.componentId ?? '']?.name ?? node.name ?? 'unnamed'
          if (!unknown.has(missing)) {
            unknown.set(missing, {
              name: missing,
              width: round(node.absoluteBoundingBox?.width),
              height: round(node.absoluteBoundingBox?.height),
              radius: round(node.cornerRadius),
            })
          }
          // Stop here. Text inside a component that will not be built is copy with nowhere to
          // go, and the issue tells the agent to use the words it is given exactly. Handing it
          // an orphaned caption produced exactly that: a stray line where the component should
          // have been. Text inside a MAPPED instance is still wanted, because that is a
          // button's label.
          return
        }
      }
      if (node.type === 'TEXT' && node.characters) text.push(node.characters)
      for (const child of node.children ?? []) visit(child)
    }
    visit(entry.document)

    const outline = outlineOf(entry.document, meta, packComponents)

    return {
      nodeId,
      name: entry.document?.name ?? 'unnamed frame',
      pageName: await this.pageOf(fileKey, nodeId),
      fileKey,
      fileName: data.name ?? '',
      components: [...known].sort(),
      // Plain codepoint order, matching Python's sorted(). localeCompare would not, and the two
      // runtimes are diffed byte for byte.
      unknownComponents: [...unknown.values()].sort((a, b) =>
        a.name < b.name ? -1 : a.name > b.name ? 1 : 0,
      ),
      text,
      outline,
    }
  }

  /** The page a node sits on. One shallow read, and a miss is not worth failing the request for. */
  private async pageOf(fileKey: string, nodeId: string): Promise<string> {
    try {
      const tree = await this.get(`/v1/files/${fileKey}`, { depth: '2' })
      for (const page of tree?.document?.children ?? []) {
        for (const child of page.children ?? []) {
          if (child.id === nodeId) return page.name ?? ''
        }
      }
    } catch {
      return ''
    }
    return ''
  }

  /**
   * A rendered PNG of the frame, as a temporary URL Figma hosts.
   *
   * The URL expires. The issue carries it anyway, because a reviewer reads the issue within
   * hours and the alternative is this service storing and serving images, which is a different
   * and much larger thing to run.
   */
  async image(fileKey: string, nodeId: string, scale = 2): Promise<string | null> {
    const data = await this.get(`/v1/images/${fileKey}`, {
      ids: nodeId,
      format: 'png',
      scale: String(scale),
    })
    return data?.images?.[nodeId] ?? null
  }

  /** Pin a link to a frame, where it shows in Dev Mode next to the design. */
  async pinDevResource(fileKey: string, nodeId: string, name: string, url: string): Promise<void> {
    await this.post('/v1/dev_resources', {
      dev_resources: [{ name, url, file_key: fileKey, node_id: nodeId }],
    })
  }

  /** Leave a comment attached to a frame. */
  async comment(fileKey: string, nodeId: string, message: string): Promise<void> {
    await this.post(`/v1/files/${fileKey}/comments`, {
      message,
      comment_pin_corner: 'top-left',
      client_meta: { node_id: nodeId, node_offset: { x: 0, y: 0 } },
    })
  }
}

File: bridge/src/bridge/figma.py
"""The Figma side of the bridge: read a frame, and pin things back onto it.

Three facts shape this module, and they are worth stating because they are the reason the
system is arranged the way it is.

The REST API can read any node and can write dev resources and comments. It cannot create or
change a single node, so nothing here can damage a designer's file. The worst this credential
can do is leave a link and a comment.

The REST API also cannot set a frame's dev status. Marking a frame *completed* stays a person's
click, which is the correct place for it: the system reports what happened and a person decides
what that means.

A frame names components by their Figma component key. The pack names the same components by
their code import. Code Connect is what joins the two, and until a component is mapped this
module can say a frame uses "some component" and not which one.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import httpx

API = "https://api.figma.com"


def _round(value: float | int | None) -> int | None:
    """Figma reports fractional sizes. Round, and treat a missing or zero value as unknown."""
    if isinstance(value, (int, float)) and value > 0:
        return round(value)
    return None


@dataclass(frozen=True)
class UnmappedComponent:
    """A component in the frame that the key map could not place.

    Carries the size the design gave it, so a placeholder can reserve the same space rather than
    collapsing the layout around a hole. None where Figma did not report one.
    """

    name: str
    width: int | None = None
    height: int | None = None
    radius: int | None = None


@dataclass(frozen=True)
class FrameRead:
    """What the bridge learned about a frame, and what it could not learn."""

    node_id: str
    name: str
    page_name: str
    file_key: str
    file_name: str
    components: list[str] = field(default_factory=list)
    unknown_components: list[UnmappedComponent] = field(default_factory=list)
    text: list[str] = field(default_factory=list)
    # How the frame is arranged, top to bottom, already rendered as markdown list lines. The
    # component list is a SET, sorted, so it says what the frame needs and nothing about where
    # anything sits. An agent handed only that has to invent an arrangement, and it did.
    outline: list[str] = field(default_factory=list)
    image_url: str | None = None

    @property
    def url(self) -> str:
        """The link a person opens. Figma writes 41-207 in a URL and 41:207 in the API."""
        node = self.node_id.replace(":", "-")
        return f"https://www.figma.com/design/{self.file_key}/?node-id={node}"


def _quoted(text: str) -> str:
    """A quoted, shortened line of copy. Long strings make an outline unreadable."""
    short = f"{text[:57]}..." if len(text) > 60 else text
    return f'"{short}"'


def _outline_of(
    root: dict,
    meta: dict,
    pack_components: dict[str, str],
    depth: int = 0,
) -> list[str]:
    """The frame's arrangement, in document order, as markdown list lines.

    Instances are opaque. The walk does not descend into one, mapped or not, because what is
    inside a component is the component's business and listing it invites an agent to rebuild it.

    A frame that lays its children out becomes a level and says which way it runs. A frame that
    does not is transparent: it nests nothing and its children are emitted where it sat, because a
    grouping with no layout is a drawing convenience rather than structure.

    Adjacent identical entries collapse to `x N`. Three figures in a row is one fact, not three.
    """
    if depth > 3:
        return []
    pad = "  " * depth
    entries: list[str] = []
    nested: dict[int, list[str]] = {}

    for child in root.get("children") or ():
        kind = child.get("type")
        if kind == "TEXT" and child.get("characters"):
            entries.append(f"{pad}- {_quoted(child['characters'])}")
            continue
        if kind == "INSTANCE":
            component_id = child.get("componentId", "")
            key = (meta.get(component_id) or {}).get("key", "")
            mapped = pack_components.get(key)
            name = (
                mapped or (meta.get(component_id) or {}).get("name") or child.get("name", "unnamed")
            )
            entries.append(f"{pad}- " + (f"`{name}`" if mapped else f"**{name}**"))
            continue
        if child.get("children"):
            layout = child.get("layoutMode")
            if layout in ("HORIZONTAL", "VERTICAL"):
                label = "a row of:" if layout == "HORIZONTAL" else "a column of:"
                entries.append(f"{pad}- {label}")
                nested[len(entries) - 1] = _outline_of(child, meta, pack_components, depth + 1)
            else:
                # Transparent. A grouping with no layout is a drawing convenience, not structure.
                entries.extend(_outline_of(child, meta, pack_components, depth))

    lines: list[str] = []
    index = 0
    while index < len(entries):
        run = 1
        while (
            index + run < len(entries)
            and entries[index + run] == entries[index]
            and index not in nested
            and index + run not in nested
        ):
            run += 1
        lines.append(f"{entries[index]} x {run}" if run > 1 else entries[index])
        lines.extend(nested.get(index, ()))
        index += run
    return lines


class FigmaClient:
    """A narrow client. Every method here is one call the bridge actually makes."""

    def __init__(self, token: str, client: httpx.Client | None = None) -> None:
        self._client = client or httpx.Client(timeout=20.0)
        self._headers = {"X-Figma-Token": token}

    def _get(self, path: str, **params: str) -> dict:
        response = self._client.get(f"{API}{path}", headers=self._headers, params=params or None)
        response.raise_for_status()
        return response.json()

    def _post(self, path: str, payload: dict) -> dict:
        response = self._client.post(f"{API}{path}", headers=self._headers, json=payload)
        response.raise_for_status()
        return response.json()

    # ── reading ────────────────────────────────────────────────────────────

    def read_frame(self, file_key: str, node_id: str, pack_components: dict[str, str]) -> FrameRead:
        """Read one frame and resolve what it is made of against the pack.

        `pack_components` maps a Figma component key to a pack component name, which is what
        Code Connect gives us. A component in the frame that is not in that map goes into
        `unknown_components` rather than being guessed at: an agent told the wrong component
        name confidently is worse than an agent told the truth vaguely.
        """
        data = self._get(f"/v1/files/{file_key}/nodes", ids=node_id, depth="8")
        entry = (data.get("nodes") or {}).get(node_id)
        if not entry:
            raise LookupError(f"node {node_id} is not in file {file_key}")

        document = entry["document"]
        components_meta = entry.get("components") or {}
        known: list[str] = []
        unknown: dict[str, UnmappedComponent] = {}
        text: list[str] = []

        def visit(node: dict) -> None:
            if node.get("type") == "INSTANCE":
                component_id = node.get("componentId", "")
                key = (components_meta.get(component_id) or {}).get("key", "")
                name = pack_components.get(key)
                if name:
                    known.append(name)
                else:
                    missing = (components_meta.get(component_id) or {}).get("name") or node.get(
                        "name", "unnamed"
                    )
                    if missing not in unknown:
                        box = node.get("absoluteBoundingBox") or {}
                        unknown[missing] = UnmappedComponent(
                            name=missing,
                            width=_round(box.get("width")),
                            height=_round(box.get("height")),
                            radius=_round(node.get("cornerRadius")),
                        )
                    # Stop here. Text inside a component that will not be built is copy with
                    # nowhere to go, and the issue tells the agent to use the words it is given
                    # exactly. Handing it an orphaned caption produced exactly that: a stray line
                    # where the component should have been. Text inside a MAPPED instance is still
                    # wanted, because that is a button's label.
                    return
            if node.get("type") == "TEXT" and node.get("characters"):
                text.append(node["characters"])
            for child in node.get("children") or ():
                visit(child)

        visit(document)

        outline = _outline_of(document, components_meta, pack_components)

        return FrameRead(
            node_id=node_id,
            name=document.get("name", "unnamed frame"),
            page_name=self._page_of(file_key, node_id),
            file_key=file_key,
            file_name=data.get("name", ""),
            components=sorted(set(known)),
            unknown_components=sorted(unknown.values(), key=lambda c: c.name),
            text=text,
            outline=outline,
        )

    def _page_of(self, file_key: str, node_id: str) -> str:
        """The page a node sits on. Cheap: one shallow read of the file's top level."""
        try:
            tree = self._get(f"/v1/files/{file_key}", depth="2")
        except httpx.HTTPError:
            return ""
        for page in (tree.get("document") or {}).get("children") or ():
            for child in page.get("children") or ():
                if child.get("id") == node_id:
                    return page.get("name", "")
        return ""

    def image(self, file_key: str, node_id: str, scale: int = 2) -> str | None:
        """A rendered PNG of the frame, as a temporary URL Figma hosts.

        The URL expires. The issue carries it anyway, because a reviewer reads the issue within
        hours and the alternative is this service storing and serving images, which is a
        different and much larger thing to run.
        """
        data = self._get(f"/v1/images/{file_key}", ids=node_id, format="png", scale=str(scale))
        return (data.get("images") or {}).get(node_id)

    # ── writing, which is only ever a link or a comment ────────────────────

    def pin_dev_resource(self, file_key: str, node_id: str, name: str, url: str) -> dict:
        """Pin a link to a frame, where it shows in Dev Mode next to the design."""
        return self._post(
            "/v1/dev_resources",
            {
                "dev_resources": [
                    {"name": name, "url": url, "file_key": file_key, "node_id": node_id}
                ]
            },
        )

    def comment(self, file_key: str, node_id: str, message: str) -> dict:
        """Leave a comment attached to a frame."""
        return self._post(
            f"/v1/files/{file_key}/comments",
            {
                "message": message,
                "comment_pin_corner": "top-left",
                "client_meta": {"node_id": node_id, "node_offset": {"x": 0, "y": 0}},
            },
        )
